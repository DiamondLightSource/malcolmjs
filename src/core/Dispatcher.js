/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
