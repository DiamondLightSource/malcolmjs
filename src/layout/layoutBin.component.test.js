import React from 'react';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import configureStore from 'redux-mock-store';
import LayoutBin from './layoutBin.component';
import {
  MalcolmInLayoutDeleteZoneType,
  MalcolmShowBinType,
} from '../malcolm/malcolm.types';
import layoutAction from '../malcolm/actions/layout.action';

jest.mock('../malcolm/actions/layout.action');

describe('Layout Bin', () => {
  let shallow;
  let mount;
  let store;

  beforeEach(() => {
    layoutAction.deleteBlocks.mockImplementation(() => ({
      type: 'test_delete_blocks',
    }));
    layoutAction.mouseInsideDeleteZone.mockImplementation(inside => ({
      type: MalcolmInLayoutDeleteZoneType,
      payload: { insideZone: inside },
    }));
    layoutAction.showLayoutBin.mockImplementation(show => ({
      type: MalcolmShowBinType,
      payload: { visible: show },
    }));

    shallow = createShallow({ dive: true });
    mount = createMount();
    const state = {};
    store = configureStore()(state);
  });

  afterEach(() => {
    mount.cleanUp();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<LayoutBin store={store} />);
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('notifies when inside delete zone', () => {
    const wrapper = mount(<LayoutBin store={store} />);

    wrapper.find('div').simulate('mouseEnter');

    const actions = store.getActions();
    expect(actions.length).toEqual(1);
    expect(actions[0].type).toEqual(MalcolmInLayoutDeleteZoneType);
    expect(actions[0].payload.insideZone).toBeTruthy();
  });

  it('notifies when leaving delete zone', () => {
    const wrapper = mount(<LayoutBin store={store} />);

    wrapper.find('div').simulate('mouseLeave');

    const actions = store.getActions();
    expect(actions.length).toEqual(1);
    expect(actions[0].type).toEqual(MalcolmInLayoutDeleteZoneType);
    expect(actions[0].payload.insideZone).toBeFalsy();
  });

  it('stops showing bin and deletes blocks when blocks are dropped on it', () => {
    const wrapper = mount(<LayoutBin store={store} />);

    wrapper.find('div').simulate('mouseUp');

    const actions = store.getActions();
    expect(actions.length).toEqual(2);
    expect(actions[0].type).toEqual(MalcolmShowBinType);
    expect(actions[0].payload.visible).toBeFalsy();

    expect(actions[1].type).toEqual('test_delete_blocks');
  });
});
