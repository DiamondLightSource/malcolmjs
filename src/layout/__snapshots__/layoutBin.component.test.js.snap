// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout Bin renders correctly 1`] = `
<div
  className="LayoutBin-container-1"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  role="presentation"
>
  <pure(Delete)
    className="LayoutBin-icon-2"
  />
</div>
`;
